File: utils/Arena.zig
const std = @import("std");
const mAlloc = std.c.malloc;
const mFree = std.c.free;
const print = std.log.info;
const CHUNK_QUANT = 32768;
const MIN_CHUNK_SIZE = (CHUNK_QUANT - 32); // 32 bytes for malloc bookkeeping

const ArenaListChunks = struct {
    payload: *[]u8,
    next: ?*ArenaListChunks,
};

const ArenaError = error {
    outOfMemory,    
};

pub const Arena = struct {
    const This = @This();
    chunksHead: ?*ArenaListChunks,
    chunksTail: ?*ArenaListChunks,
    currChunk: *[]u8,
    currInd: u32,

    pub fn allocate(this: This, T: type, allocSize: usize) ArenaError!T {
        if (this.currInd + allocSize >= this.currChunk.len) {
            const newSize = calculateChunkSize(allocSize);
            const newChunk: *[newSize]u8 = try mAlloc(newSize + @sizeOf(ArenaListChunks));
            const newLink = @ptrCast(*ArenaListChunks, newChunk + newSize);
            this.chunksTail.next = newLink;
            this.chunksTail = newLink;
            print("Allocated a new chunk\n", .{});
            this.currChunk = newChunk;
            this.currInd = 0;
            
        }
        const result = @ptrCast(T, &this.currChunk[this.currInd]);
        this.currInd += allocSize;
        return result;
    }

    pub fn delete(this: *This) void {
        var curr = this.chunksHead;
        //var nextToFree = curr;
        while (curr != null) {
            const curr2 = curr.?;
            var nextToFree: ?*ArenaListChunks = curr2.next;
            print("freeing a chunk\n", .{});
            
            mFree(curr2.payload);
            curr = nextToFree;
        }
        print("Freeing arena head", .{});
        mFree(this);
    }
};

fn calculateChunkSize(allocSize: usize) usize {
    const fullMemory = @sizeOf(ArenaListChunks) + allocSize + 32;
    const mallocMemory = if (fullMemory < CHUNK_QUANT)
                            { CHUNK_QUANT; }
                            else {
                                if (fullMemory % CHUNK_QUANT > 0) 
                                    {(fullMemory/CHUNK_QUANT + 1)*CHUNK_QUANT;}
                                    else fullMemory;
                            };
    return mallocMemory - 32;    
}

fn rawAlloc(comptime T: type, sz: u32) ArenaError!*T {
    const result = @ptrCast(?*T, @alignCast(@alignOf(*u64), mAlloc(sz)));
    if (result == null) return ArenaError.outOfMemory;
    return result.?;
}

pub fn mkArena() ArenaError!*Arena {    
    const result = try rawAlloc(Arena, @sizeOf(Arena));    

    const firstChunk: *[]u8 = try rawAlloc([]u8, MIN_CHUNK_SIZE + @sizeOf(ArenaListChunks));
    const listHead = @ptrCast(*ArenaListChunks, firstChunk + @ptrCast(*u8, MIN_CHUNK_SIZE));
    result.chunksHead = listHead;
    result.chunksTail = listHead;
    result.chunksHead.payload = firstChunk;
    result.chunksHead.next = null;
    result.currChunk = firstChunk;
    result.currInd = 0;
    return result;
}

File: utils/Arena.c
#include <stdio.h>
#include <stdlib.h>
#include "Arena.h"

#define CHUNK_QUANT 32768


struct ArenaChunk {
    size_t size;
    ArenaChunk* next;
    char memory[]; // flexible array member
};

struct Arena {
    ArenaChunk* firstChunk;
    ArenaChunk* currChunk;
    int currInd;
};


size_t minChunkSize() {
    return (size_t)(CHUNK_QUANT - 32);
}

Arena* mkArena() {
    Arena* result = malloc(sizeof(Arena));

    size_t firstChunkSize = minChunkSize();

    ArenaChunk* firstChunk = malloc(firstChunkSize);

    firstChunk->size = firstChunkSize - sizeof(ArenaChunk);
    firstChunk->next = NULL;

    result->firstChunk = firstChunk;
    result->currChunk = firstChunk;
    result->currInd = 0;

    return result;
}

/**
 * Calculates memory for a new chunk. Memory is quantized and is always 32 bytes less
 */
size_t calculateChunkSize(size_t allocSize) {
    // 32 for any possible padding malloc might use internally,
    // so that the total allocation size is a good even number of OS memory pages.
    // TODO review
    size_t fullMemory = sizeof(ArenaChunk) + allocSize + 32; // struct header + main memory chunk + space for malloc bookkeep
    int mallocMemory = fullMemory < CHUNK_QUANT
                        ? CHUNK_QUANT
                        : (fullMemory % CHUNK_QUANT > 0
                           ? (fullMemory/CHUNK_QUANT + 1)*CHUNK_QUANT
                           : fullMemory);

    return mallocMemory - 32;
}

/**
 * Allocate memory in the arena, malloc'ing a new chunk if needed
 */
void* arenaAllocate(Arena* ar, size_t allocSize) {
    if (ar->currInd + allocSize >= ar->currChunk->size) {
        size_t newSize = calculateChunkSize(allocSize);

        ArenaChunk* newChunk = malloc(newSize);
        if (!newChunk) {
            perror("malloc error when allocating arena chunk");
            exit(EXIT_FAILURE);
        };
        // sizeof includes everything but the flexible array member, that's why we subtract it
        newChunk->size = newSize - sizeof(ArenaChunk);
        newChunk->next = NULL;
        printf("Allocated a new chunk with bookkeep size %zu, array size %zu\n", sizeof(ArenaChunk), newChunk->size);

        ar->currChunk->next = newChunk;
        ar->currChunk = newChunk;
        ar->currInd = 0;
    }
    void* result = (void*)(ar->currChunk->memory + (ar->currInd));
    ar->currInd += allocSize;
    return result;
}


void arenaDelete(Arena* ar) {
    ArenaChunk* curr = ar->firstChunk;
    ArenaChunk* nextToFree = curr;
    while (curr != NULL) {
        printf("freeing a chunk of size %zu\n", curr->size);
        nextToFree = curr->next;
        free(curr);
        curr = nextToFree;
    }
    printf("freeing arena itself\n");
    free(ar);
}
